pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,663 @@
+        Ok(tenant) => tenant,
+        Err(e) => {
+            // this can happen if tenant has lifecycle transition after we fetched it
+            debug!("failed to get tenant: {e:#}");
+            return ControlFlow::Continue(());
+        }
+    };
+
+    if cancel.is_cancelled() {
+        return ControlFlow::Break(());
+    }
+
+    let mut max_layer_size = ApproxAccurate::default();
+    for tl in tenant.list_timelines() {
+        if !tl.is_active() {
+            continue;
+        }
+        let info = tl.get_local_layers_for_disk_usage_eviction();
+        debug!(timeline_id=%tl.timeline_id, "timeline resident layers count: {}", info.resident_layers.len());
+        scratch.extend(
+            info.resident_layers
+                .into_iter()
+                .map(|layer_infos| (tl.clone(), layer_infos)),
+        );
+        max_layer_size = max_layer_size.max(info.max_layer_size.accurate());
+
+        if cancel.is_cancelled() {
+            return ControlFlow::Break(());
+        }
+    }
+
+    let min_resident_size = match mode {
+        Mode::GlobalLru => {
+            lru_candidates.append(scratch);
+            return ControlFlow::Continue(());
+        }
+        Mode::RespectTenantMinResidentSize => match tenant.get_min_resident_size_override() {
+            Some(size) => size,
+            None => {
+                match max_layer_size.accurate() {
+                    Some(size) => size,
+                    None => {
+                        let prod_max_layer_file_size = 332_880_000;
+                        // rate-limit warning in case above comment is wrong and we're missing `LayerMetadata` for many layers
+                        static LAST_WARNED: Mutex<Option<Instant>> = Mutex::new(None);
+                        let mut last_warned = LAST_WARNED.lock().unwrap();
+                        if last_warned
+                            .map(|v| v.elapsed() > Duration::from_secs(60))
+                            .unwrap_or(true)
+                        {
+                            warn!(value=prod_max_layer_file_size, "some layers don't have LayerMetadata to calculate max_layer_file_size, using default value");
+                            *last_warned = Some(Instant::now());
+                        }
+                        prod_max_layer_file_size
+                    }
+                }
+            }
+        },
+    };
+
+    scratch.sort_unstable_by_key(|(_, layer_info)| layer_info.last_activity_ts);
+
+    let mut current: u64 = scratch.iter().map(|(_, layer)| layer.file_size()).sum();
+    for (tl, layer) in scratch.drain(..) {
+        if cancel.is_cancelled() {
+            return ControlFlow::Break(());
+        }
+        if current <= min_resident_size {
+            break;
+        }
+        current -= layer.file_size();
+        debug!(?layer, "adding layer to lru_candidates");
+        lru_candidates.push((tl, layer));
+    }
```

Comment:
This is "off by one" compared to what I expected. For example, imagine that there are five layers, each one has size 10.

```
     layer size    cumulative size
A:   10            50
B:   10            40
C:   10            30
D:   10            20
E:   10            10
```
Because `min_resident_size' is 25, I would expect this to "protect" evict layers A-B, and protect layers C-E. Not so! This will in fact evict layers A-C, protecting only D-E.

This looks worse if the layers have very different sizes. For example:

```
     layer size    cumulative size
A:   10            47
B:   10            37
C:   25            27
D:    1             2
E:    1             1
```
If `min_resident_size` is again 25, this again protects only D-E, choosing to evict the large C layer.

I'm not sure if this was intentional or not, but I was surprised. I expected `min_resident_size` to mean "keep at least this much data for every tenant, if possible", but as it stands, it doesn't quite do that.

Reply:
> ```
>      layer size    cumulative size
> A:   10            47
> B:   10            37
> C:   25            27
> D:    1             2
> E:    1             1
> ```
>
> If `min_resident_size` is again 25, this again protects only D-E, choosing to evict the large C layer.
>
> I'm not sure if this was intentional or not, but I was surprised. I expected `min_resident_size` to mean "keep at least this much data for every tenant, if possible", but as it stands, it doesn't quite do that.

Could you reiterate how does it not do what you expected? I think you just wrote that it does exactly what you were expecting ("if possible" part).

By design, most recently used layers should be spared, and reading through this example perhaps instead of rewriting a change like:

```
if current - layer.file_size() <= min_resident_size {
```

Would had made it work 100% how you expected?

Reply:
> > ```
> >      layer size    cumulative size
> > A:   10            47
> > B:   10            37
> > C:   25            27
> > D:    1             2
> > E:    1             1
> > ```
> > 
> > If `min_resident_size` is again 25, this again protects only D-E, choosing to evict the large C layer.
> > I'm not sure if this was intentional or not, but I was surprised. I expected `min_resident_size` to mean "keep at least this much data for every tenant, if possible", but as it stands, it doesn't quite do that.
> 
> Could you reiterate how does it not do what you expected? I think you just wrote that it does exactly what you were expecting ("if possible" part).

`min_resident_size` is 25, so I expected it to retain at least 25 (bytes) worth of layers. but it only retained 2 (D and E, total size 2). I expected it to retain layers C, D and E, they would have total size of 27. 27 would be greater than `min_resident_size`, but 2 is not.

> By design, most recently used layers should be spared, and reading through this example perhaps instead of a full rewrite, a change like below would had made it work 100% how you expected:
> 
> ```
> if current - layer.file_size() <= min_resident_size {
> ```

Sounds correct. (Would need another cup of coffee to verify :-) )

Reply:
As for a single layer at the boundary being included or excluded, I don't find that interesting because we will on-demand download a missing layer regardless. However, keeping the tiny quite recent layers might be beneficial to avoid the churn.

Besides, there is no way for this to be fully accurate; we don't subscribe to the individual layer access times getting updated nor have a check while evicting that we are in fact, evicting a expectedly-least-recently-used layer. And I think it's fine, this is a best effort tool which will hopefully keep lockups from happening at a some low cost.

Reply:
I agree with that sentiment

Reply:
Rewrite has been merged which has the behaviour you wanted.